plutus-core/plutus-core/test/Evaluation/Spec.hs

```diff
@@ -1,11 +1,247 @@
+  let (res, logs) = runEmitter . runExceptT $ eval args argStrings
+  case res of
+    Right _ -> success
+    Left err -> do
+      liftIO $ do
+        putStrLn "Builtin function evaluation failed"
+        putStrLn $ "Function: " <> show bn
+        putStrLn $ "Arguments: " <> show argStrings
+        putStrLn $
+          "Error: "
+            <> ( case err of
+                   KnownTypeEvaluationFailure -> "KnownTypeEvaluationFailure"
+                   KnownTypeUnliftingError e ->
+                     "KnownTypeUnliftingError: "
+                       <> show (pretty e)
+               )
+        putStrLn $ "Execution log: " <> show logs
+      failure
+  where
+    meaning :: BuiltinMeaning Term (CostingPart DefaultUni DefaultFun)
+    meaning = toBuiltinMeaning bn
+
+    eval :: [Term] -> [String] -> MakeKnownM Term
+    eval args0 argStrings = case meaning of
+      BuiltinMeaning _ _ runtime -> go (_broRuntimeScheme runtime) (_broImmediateF runtime) args0
+      where
+        go ::
+          forall n.
+          RuntimeScheme n ->
+          ToRuntimeDenotationType Term n ->
+          [Term] ->
+          MakeKnownM Term
+        go sch f args = case (sch, args) of
+          (RuntimeSchemeArrow sch', a : as) -> do
+            res <- liftEither (f a)
+            go sch' res as
+          (RuntimeSchemeResult, []) -> f
+          (RuntimeSchemeAll sch', _) -> go sch' f args
+          _ -> error $ "Wrong number of args for builtin " <> show bn <> ": " <> show argStrings
+
+-- | Generate arguments to a builtin function
+genArgs :: DefaultFun -> Gen [(Term, String)]
+genArgs bn = case bn of
+  -- These functions are partial, so we manually generate their arguments.
+  IndexByteString -> do
+    s <- Gen.utf8 (Range.linear 5 100) Gen.enumBounded
+    i :: Integer <- fromIntegral <$> Gen.int (Range.linear 0 (BS.length s - 1))
+    pure [mkArg s, mkArg i]
+  HeadList -> pure . pure . mkArg =<< Gen.list (Range.linear 1 100) genInteger
+  TailList -> genArgs HeadList
+  UnConstrData -> pure . pure . mkArg =<< genConstr 5
+  UnMapData -> pure . pure . mkArg =<< genMap 5
+  UnListData -> pure . pure . mkArg =<< genList 5
+  UnIData -> pure . pure . mkArg =<< genI
+  UnBData -> pure . pure . mkArg =<< genB
+  -- The rest are total functions and we generate the arguments based on their `TypeScheme`s.
+  _ -> sequenceA $ case meaning of
+    BuiltinMeaning tySch _ _ -> go tySch
+      where
+        go :: forall args res. TypeScheme Term args res -> [Gen (Term, String)]
+        go = \case
+          TypeSchemeResult    -> []
+          TypeSchemeArrow sch -> genArg (typeRep @(Head args)) : go sch
+          TypeSchemeAll _ sch -> go sch
+  where
+    meaning :: BuiltinMeaning Term (CostingPart DefaultUni DefaultFun)
+    meaning = toBuiltinMeaning bn
+
+-- | Generate one argument to a builtin function based on its `TypeRep`.
+genArg :: forall k (a :: k). TypeRep a -> Gen (Term, String)
```

Comment:
Oh, I'd use a class, but this is even better indeed. Thanks!

What about making it

```haskell
genArg :: forall k (a :: k). (DefaultUni `Contains` a, PrettyConst a) => TypeRep a -> Gen a
```

so that you don't need to wrap the resulting value with `Constant . someValue` just to unwrap it back (and handle the impossible failure) in the polymorphic cases?

Would that also allow you to get rid of `unsafeCoerce`?

Reply:
This seems like a useful function, maybe we should put it in the testlib?

Reply:
The fancy next step would be to generalize with `Everywhere uni Arbitrary`, but not sure if that's even helpful.

Reply:
Also: not clear from the doc what the string in the returned tuple is. I'm also not entirely sure why we need it - can't we just print the constant?

Reply:
> `genArg :: forall k (a :: k). (DefaultUni `Contains` a, PrettyConst a) => TypeRep a -> Gen a`

I doubt this works. First, matching on `TypeSchemeArrow` gives us `Typeable a` and a few other instances, but not `Contains` or `PrettyConst`. Second, there's no `Contains` or `PrettyConst` for `Opaque x y` and `SomeConstant x y`, right?

Reply:
> Second, there's no `Contains` or `PrettyConst` for `Opaque x y` and `SomeConstant x y`, right?

Oh, yes, right, I got it all wrong, sorry. I think there's a bigger problem here, I'll write something up a bit later.

Reply:
> I'm also not entirely sure why we need it - can't we just print the constant?

I wanted to print the arguments when the test fails. Is there a way to print an arbitrary `Some (ValueOf DefaultUni)`?

Reply:
Yup:

```
>>> import PlutusCore.Pretty
>>> :type pretty (undefined :: Some (ValueOf DefaultUni))
pretty (undefined :: Some (ValueOf DefaultUni)) :: Doc ann
```

Reply:
Never mind, there's apparently a `Pretty` instance for `Term`.